avocado/core/loader.py

```diff
@@ -1027,6 +1014,68 @@ def get_decorator_mapping():
         return {test.ExternalRunnerTest: output.TERM_SUPPORT.healthy_str}
 
 
+class PythonUnittestLoader(FileLoader):
+    name = "py_unittest"
+
+    __not_test_str = ("Does not look like a python unittest")
+
+    @staticmethod
+    def get_type_label_mapping():
+        return {NotATest: 'NOT_A_TEST',
+                MissingTest: 'MISSING',
+                BrokenSymlink: 'BROKEN_SYMLINK',
+                AccessDeniedPath: 'ACCESS_DENIED',
+                test.Test: 'INSTRUMENTED',
+                test.PythonUnittest: 'PyUNITTEST'}
+
+    @staticmethod
+    def get_decorator_mapping():
+        return {NotATest: output.TERM_SUPPORT.warn_header_str,
+                MissingTest: output.TERM_SUPPORT.fail_header_str,
+                BrokenSymlink: output.TERM_SUPPORT.fail_header_str,
+                AccessDeniedPath: output.TERM_SUPPORT.fail_header_str,
+                test.PythonUnittest: output.TERM_SUPPORT.healthy_str}
+
+    def _make_tests(self, test_path, list_non_tests, subtests_filter=None):
+        def ignore_broken(klass, uid, description=None):
+            """ Always return empty list """
+            return []
+
+        if list_non_tests:  # return broken test with params
+            make_broken = self._make_test
+        else:  # return empty set instead
+            make_broken = ignore_broken
+        try:
+            # Python unittests
+            assert (test_path.endswith(".py") or test_path.endswith(".pyc") or
+                    test_path.endswith(".pyo"))
```

Comment:
Why do you care that much about extensions. Linux don't and users are used to it. A file with any extension can be a unittest. Or without an extension. Why to limit that?

Reply:
As I mentioned earlier, the `python -m unittest` __ONLY__ discovers those extensions, no matter what users think about it. And as safeloader only parses `.py` files I think we should even simplify that and look for `.py` files only.

Reply:
I don't see where `python -m unittest` limits us. The only limitation we have is the fact that safeloader needs to read the file, so it has to be not-compiled. Nevermind, it's not important. I just see as meaningless caring that much about file extensions.

Reply:
I'd say requiring the files to be .py files respects the principle of least surprise. Even if linux/unix in general does not care terribly about extensions, the fact that the python unittest loader itself adopts this behavior is telling: We would like users to follow certain good practices. Having a .sh file to contain python code or a .py file containing shell code leads to confusion, so honoring file extensions is justifiable, since it keeps things simple and easy to follow.

Reply:
```bash
$ ls
test.py
$ python -m unittest test  # works well
$ ls
test.pyc
$ python -m unittest test  # works well but is not discoverable by safeloader
$ ls
test
$ python -m unittest test  # fails as test does not exists
$ ls
test.py.foo
$ python -m unittest test  # fails as test does not exists
$ python -m unittest test.py  # fails as test does not exists
$ python -m unittest test.py.foo  # fails as test does not exists
```

Reply:
`.sh` with python code? I have never mentioned that use case, but it should work:

```
/tmp$ echo 'print("I am python!")' > test.sh
/tmp$ python test.sh
I am python!
```
Guys, we are comparing avocados with apples and trying to paint the avocados red so they look more like apples, despite the taste. unittest can find tests where Avocado can't (`pyc`) because they don't really load the tests safely. On the other hand, Avocado loads the tests safely and **could** find tests where unittest can't (any non-binary file, regardless the extension), but we prefer to limit our scope to filenames terminated in `.py` so we don't "suprise" users when they provide an extension-less file (or a file with any extension) and it works. I know that breaking that limitation would have other implications, like when users provide a directory, not a file, as we currently also white-list the extensions we look for in a directory. As I said, I think it's a meaningless limitation, but its existence is not a blocker.  Let's move on.

Regarding this specific line of code, you can drop the `pyc`/`pyo`, as safeloader will not load those anyway.